sql: Panic after moving table from REGIONAL BY ROW to REGIONAL BY TABLE and back to REGIONAL BY ROW

When running the included script, ALTERs from REGIONAL BY ROW to REGIONAL BY TABLE, and back to REGIONAL BY ROW will succeed, but then eventually panic with the below stack.  Hit this on `ajstorm_zone_config_force_table` branch, but it might also be reproducible on master.

[zone_configs_repro.scr.zip](https://github.com/cockroachdb/cockroach/files/6111844/zone_configs_repro.scr.zip)


```
* ERROR: a panic has occurred!
* use of enum metadata before hydration as an enum
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	/usr/local/opt/go/libexec/src/runtime/panic.go:969
*   | [...repeated from below...]
* Wraps: (2) assertion failure
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql/types.(*T).ensureHydratedEnum
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:2415
*   | github.com/cockroachdb/cockroach/pkg/sql/types.(*T).EnumGetIdxOfPhysical
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:2381
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.GetEnumComponentsFromPhysicalRep
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:3977
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.DecodeUntaggedDatum
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/column_type_encoding.go:674
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.DecodeTableValue
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/column_type_encoding.go:524
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.DecodePartitionTuple
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/partition.go:146
*   | github.com/cockroachdb/cockroach/pkg/sql.indexCoveringsForPartitioning
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/partition_utils.go:202
*   | github.com/cockroachdb/cockroach/pkg/sql.GenerateSubzoneSpans.func1
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/partition_utils.go:118
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog.ForEachIndex
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descriptor.go:613
*   | github.com/cockroachdb/cockroach/pkg/sql.GenerateSubzoneSpans
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/partition_utils.go:103
*   | github.com/cockroachdb/cockroach/pkg/sql.writeZoneConfig
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_zone_config.go:903
*   | github.com/cockroachdb/cockroach/pkg/sql.RemoveIndexZoneConfigs
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_zone_config.go:989
*   | github.com/cockroachdb/cockroach/pkg/sql/gcjob.gcIndexes.func2
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/index_garbage_collection.go:69
*   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:772
*   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:828
*   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:771
*   | github.com/cockroachdb/cockroach/pkg/sql/gcjob.gcIndexes
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/index_garbage_collection.go:68
*   | github.com/cockroachdb/cockroach/pkg/sql/gcjob.performGC
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:66
*   | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:158
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1178
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1179
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:255
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:214
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1
*   | 	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351
*   | runtime.goexit
*   | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1374
* Wraps: (4) use of enum metadata before hydration as an enum
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
panic: use of enum metadata before hydration as an enum [recovered]
	panic: use of enum metadata before hydration as an enum

goroutine 48849 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000557d80, 0x943d0a0, 0xc0015362c0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:233 +0x126
panic(0x8196600, 0xc014f7ac80)
	/usr/local/opt/go/libexec/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).ensureHydratedEnum(...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:2415
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).EnumGetIdxOfPhysical(0xc0109f7680, 0xc006d6434e, 0x1, 0x1, 0x1, 0x0, 0x0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:2381 +0x426
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.GetEnumComponentsFromPhysicalRep(0xc0109f7680, 0xc006d6434e, 0x1, 0x1, 0x0, 0x0, 0xc006d6434e, 0x1, 0x1, 0x0, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:3977 +0x4d
github.com/cockroachdb/cockroach/pkg/sql/rowenc.DecodeUntaggedDatum(0xc00768ce10, 0xc0109f7680, 0xc006d6434d, 0x2, 0x2, 0x0, 0x6, 0x0, 0x0, 0x2, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/column_type_encoding.go:674 +0x3385
github.com/cockroachdb/cockroach/pkg/sql/rowenc.DecodeTableValue(0xc00768ce10, 0xc0109f7680, 0xc006d6434c, 0x3, 0x3, 0x0, 0x6, 0x0, 0x0, 0x8, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/column_type_encoding.go:524 +0x165
github.com/cockroachdb/cockroach/pkg/sql/rowenc.DecodePartitionTuple(0xc00768ce10, 0xc0017be500, 0xc0017be500, 0x0, 0x95ac880, 0xc0012c7200, 0xc0012c7288, 0xc0012c73c8, 0xc006d6434c, 0x3, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/partition.go:146 +0x492
github.com/cockroachdb/cockroach/pkg/sql.indexCoveringsForPartitioning(0xc00768ce10, 0xc0017be500, 0xc0017be500, 0x0, 0x95ac880, 0xc0012c7200, 0xc0012c7288, 0xc0012c73c8, 0xc00768c8b0, 0x0, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/partition_utils.go:202 +0xb3f
github.com/cockroachdb/cockroach/pkg/sql.GenerateSubzoneSpans.func1(0x959dc60, 0xc0120a5c40, 0x10000, 0x959dc60)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/partition_utils.go:118 +0x1ae
github.com/cockroachdb/cockroach/pkg/sql/catalog.ForEachIndex(0x95ac880, 0xc0012c7200, 0xc006010000, 0xc00768c9c8, 0xc00768cab0, 0xc00de56800)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descriptor.go:613 +0x11f
github.com/cockroachdb/cockroach/pkg/sql.GenerateSubzoneSpans(0xc0007ed500, 0xff49ba3b6e8caeb5, 0x6f341dfbc6c164a5, 0xc0017be500, 0xc0017be500, 0x0, 0x95ac880, 0xc0012c7200, 0xc005dc7c00, 0x6, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/partition_utils.go:103 +0x375
github.com/cockroachdb/cockroach/pkg/sql.writeZoneConfig(0x943d160, 0xc011f93290, 0xc0036ff830, 0xc000000040, 0x95ac880, 0xc0012c7200, 0xc0109f78c0, 0xc001594000, 0x0, 0x0, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_zone_config.go:903 +0x498
github.com/cockroachdb/cockroach/pkg/sql.RemoveIndexZoneConfigs(0x943d160, 0xc011f93290, 0xc0036ff830, 0xc001594000, 0x40, 0xc00768d670, 0x1, 0x1, 0x0, 0xc0007ed500)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_zone_config.go:989 +0x2d5
github.com/cockroachdb/cockroach/pkg/sql/gcjob.gcIndexes.func2(0x943d160, 0xc011f93290, 0xc0036ff830, 0x0, 0x0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/index_garbage_collection.go:69 +0x111
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x943d160, 0xc011f93290, 0xc0036ff830, 0x3ff0000000000000, 0x9591000)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:772 +0x43
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc0036ff830, 0x943d160, 0xc011f93290, 0xc00768dbe0, 0xc0036ff830, 0x7ebfc00)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:828 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc0010ceb60, 0x943d160, 0xc011f93290, 0xc00768dee0, 0xc00781c900, 0x0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:771 +0x105
github.com/cockroachdb/cockroach/pkg/sql/gcjob.gcIndexes(0x943d160, 0xc011f93290, 0xc001594000, 0x40, 0xc0174b5d00, 0xc0100ff550, 0x76170c0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/index_garbage_collection.go:68 +0x31a
github.com/cockroachdb/cockroach/pkg/sql/gcjob.performGC(0x943d160, 0xc011f93290, 0xc001594000, 0xc01889ecc0, 0xc0174b5d00, 0xc00a3a6740, 0x28)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:66 +0x225
github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume(0x8e118e573060001, 0x943d160, 0xc011f93290, 0x82a5600, 0xc004c84ee0, 0x0, 0x0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:158 +0x578
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1(0xc00159e0c0, 0x93e7720, 0xc00ac46af8, 0x943d160, 0xc011f93290, 0x82a5600, 0xc004c84ee0, 0xc00768fa50)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1178 +0xae
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc000789040, 0x943d160, 0xc011f93230, 0x82a5600, 0xc004c84ee0, 0x93e7720, 0xc00ac46af8, 0xc0126f72c0, 0xc002526901, 0x7, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1179 +0x885
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc000789040, 0x943d0a0, 0xc018c879c0, 0x93e7720, 0xc00ac46af8, 0xc0126f72c0, 0xc002526901, 0x7, 0xc00d7be540, 0x16, ...)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:255 +0x378
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1(0x943d0a0, 0xc0015362c0)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:214 +0x165
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc000557d80, 0x943d0a0, 0xc0015362c0, 0x0, 0xc012fafb30)
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351 +0xb9
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/Users/storm/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0xfc```

--- a/pkg/ccl/multiregionccl/regional_by_row_test.go
+++ b/pkg/ccl/multiregionccl/regional_by_row_test.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -549,4 +550,121 @@ COMMIT;`)
 	COMMIT;`)
 		return err
 	})
+}
+
+// TestIndexCleanupAfterAlterFromRegionalByRow ensures that old indexes for
+// REGIONAL BY ROW transitions get cleaned up correctly.
+func TestIndexCleanupAfterAlterFromRegionalByRow(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// TODO(ajstorm): Enable once #61944 goes in.
+	skip.UnderRace(t, `This test is too heavyweight to be stressed under race`)
+
+	// Decrease the adopt loop interval so that retries happen quickly.
+	defer sqltestutils.SetTestJobsAdoptInterval()()
+
+	_, sqlDB, cleanup := multiregionccltestutils.TestingCreateMultiRegionCluster(
+		t, 3 /* numServers */, base.TestingKnobs{}, nil, /* baseDir */
+	)
+	defer cleanup()
+
+	_, err := sqlDB.Exec(
+		`CREATE DATABASE "mr-zone-configs" WITH PRIMARY REGION "us-east1" REGIONS "us-east2", "us-east3";
+USE "mr-zone-configs";
+CREATE TABLE regional_by_row (
+  pk INT PRIMARY KEY,
+  i INT,
+  INDEX(i),
+  FAMILY (pk, i)
+) LOCALITY REGIONAL BY ROW`)
+	require.NoError(t, err)
+
+	// Alter the table to REGIONAL BY TABLE, and then back to REGIONAL BY ROW, to
+	// create some indexes that need cleaning up.
+	_, err = sqlDB.Exec(`ALTER TABLE regional_by_row SET LOCALITY REGIONAL BY TABLE;
+ALTER TABLE regional_by_row SET LOCALITY REGIONAL BY ROW`)
+	require.NoError(t, err)
+
+	// Validate that the indexes requiring cleanup exist.
+	type row struct {
+		status  string
+		details string
+	}
+
+	for {
+		// First confirm that the schema change job has completed
+		res := sqlDB.QueryRow(`WITH jobs AS (
+      SELECT status, crdb_internal.pb_to_json(
+			'cockroach.sql.jobs.jobspb.Payload',
+			payload,
+			false
+		) AS job
+		FROM system.jobs
+		)
+    SELECT count(*)
+    FROM jobs
+    WHERE (job->>'schemaChange') IS NOT NULL AND status = 'running'`)
+
+		require.NoError(t, res.Err())
+
+		numJobs := 0
+		err = res.Scan(&numJobs)
+		require.NoError(t, err)
+		if numJobs == 0 {
+			break
+		}
+	}
+
+	queryIndexGCJobsAndValidateCount := func(status string, expectedCount int) error {
+		query := `WITH jobs AS (
+      SELECT status, crdb_internal.pb_to_json(
+			'cockroach.sql.jobs.jobspb.Payload',
+			payload,
+			false
+		) AS job
+		FROM system.jobs
+		)
+    SELECT status, job->'schemaChangeGC' as details
+    FROM jobs
+    WHERE (job->>'schemaChangeGC') IS NOT NULL AND status = '%s'`
+
+		res, err := sqlDB.Query(fmt.Sprintf(query, status))
+		require.NoError(t, err)
+
+		var rows []row
+		for res.Next() {
+			r := row{}
+			err = res.Scan(&r.status, &r.details)
+			require.NoError(t, err)
+			rows = append(rows, r)
+		}
+		actualCount := len(rows)
+		if actualCount != expectedCount {
+			return errors.Newf("expected %d jobs with status %q, found %d. Jobs found: %v",
+				expectedCount,
+				status,
+				actualCount,
+				rows)
+		}
+		return nil
+	}
+
+	// Now check that we have the right number of index GC jobs pending.
+	err = queryIndexGCJobsAndValidateCount(`running`, 4)
+	require.NoError(t, err)
+	err = queryIndexGCJobsAndValidateCount(`succeeded`, 0)
+	require.NoError(t, err)
+
+	// Change gc.ttlseconds to speed up the cleanup.
+	_, err = sqlDB.Exec(`ALTER TABLE regional_by_row CONFIGURE ZONE USING gc.ttlseconds = 1`)
+	require.NoError(t, err)
+
+	// Validate that indexes are cleaned up.
+	queryAndEnsureThatFourIndexGCJobsSucceeded := func() error {
+		return queryIndexGCJobsAndValidateCount(`succeeded`, 4)
+	}
+	testutils.SucceedsSoon(t, queryAndEnsureThatFourIndexGCJobsSucceeded)
+	err = queryIndexGCJobsAndValidateCount(`running`, 0)
+	require.NoError(t, err)
 }

--- a/pkg/sql/region_util.go
+++ b/pkg/sql/region_util.go
@@ -453,6 +453,26 @@ func applyZoneConfigForMultiRegionTableOptionNewIndexes(
 	}
 }
 
+// dropZoneConfigsForMultiRegionIndexes drops the zone configs for all
+// the indexes defined on a multi-region table. This function is used to clean
+// up zone configs when transitioning from REGIONAL BY ROW.
+func dropZoneConfigsForMultiRegionIndexes(
+	indexIDs ...descpb.IndexID,
+) applyZoneConfigForMultiRegionTableOption {
+	return func(
+		zoneConfig zonepb.ZoneConfig,
+		regionConfig descpb.DatabaseDescriptor_RegionConfig,
+		table catalog.TableDescriptor,
+	) (hasNewSubzones bool, newZoneConfig zonepb.ZoneConfig, err error) {
+		for _, indexID := range indexIDs {
+			for _, region := range regionConfig.Regions {
+				zoneConfig.DeleteSubzone(uint32(indexID), string(region.Name))
+			}
+		}
+		return false, zoneConfig, nil
+	}
+}
+
 // isPlaceholderZoneConfigForMultiRegion returns whether a given zone config
 // should be marked as a placeholder config for a multi-region object.
 // See zonepb.IsSubzonePlaceholder for why this is necessary.
@@ -534,19 +554,19 @@ func ApplyZoneConfigForMultiRegionTable(
 	opts ...applyZoneConfigForMultiRegionTableOption,
 ) error {
 	tableID := table.GetID()
-	zoneRaw, err := getZoneConfigRaw(ctx, txn, execCfg.Codec, tableID)
+	currentZoneConfig, err := getZoneConfigRaw(ctx, txn, execCfg.Codec, tableID)
 	if err != nil {
 		return err
 	}
-	zoneConfig := *zonepb.NewZoneConfig()
-	if zoneRaw != nil {
-		zoneConfig = *zoneRaw
+	newZoneConfig := *zonepb.NewZoneConfig()
+	if currentZoneConfig != nil {
+		newZoneConfig = *currentZoneConfig
 	}
 
 	var hasNewSubzones bool
 	for _, opt := range opts {
-		newHasNewSubzones, newZoneConfig, err := opt(
-			zoneConfig,
+		newHasNewSubzones, modifiedNewZoneConfig, err := opt(
+			newZoneConfig,
 			regionConfig,
 			table,
 		)
@@ -554,26 +574,44 @@ func ApplyZoneConfigForMultiRegionTable(
 			return err
 		}
 		hasNewSubzones = newHasNewSubzones || hasNewSubzones
-		zoneConfig = newZoneConfig
+		newZoneConfig = modifiedNewZoneConfig
 	}
 
 	// Mark the NumReplicas as 0 if we have subzones but no other features
 	// in the zone config. This signifies a placeholder.
 	// Note we do not use hasNewSubzones here as there may be existing subzones
 	// on the zone config which may still be a placeholder.
-	if len(zoneConfig.Subzones) > 0 && isPlaceholderZoneConfigForMultiRegion(zoneConfig) {
-		zoneConfig.NumReplicas = proto.Int32(0)
+	if len(newZoneConfig.Subzones) > 0 && isPlaceholderZoneConfigForMultiRegion(newZoneConfig) {
+		newZoneConfig.NumReplicas = proto.Int32(0)
 	}
 
-	if !zoneConfig.Equal(zonepb.NewZoneConfig()) {
-		if err := zoneConfig.Validate(); err != nil {
+	// Determine if we're rewriting or deleting the zone configuration.
+	newZoneConfigIsEmpty := newZoneConfig.Equal(zonepb.NewZoneConfig())
+	currentZoneConfigIsEmpty := currentZoneConfig.Equal(zonepb.NewZoneConfig())
+	rewriteZoneConfig := !newZoneConfigIsEmpty
+	deleteZoneConfig := newZoneConfigIsEmpty && !currentZoneConfigIsEmpty
+
+	// It's possible at this point that we'll have an empty zone configuration
+	// that doesn't look like an empty zone configuration (i.e. it's a placeholder
+	// zone config - NumReplicas = 0 - with no subzones set). This can happen if
+	// we're ALTERing from REGIONAL BY ROW and dropping all of the subzones for
+	// the partitions. If we encounter this, instead of re-writing the zone
+	// configuration, we want to delete it.
+	numReplicasIsZero := newZoneConfig.NumReplicas != nil && *newZoneConfig.NumReplicas == 0
+	if len(newZoneConfig.Subzones) == 0 && numReplicasIsZero {
+		rewriteZoneConfig = false
+		deleteZoneConfig = true
+	}
+
+	if rewriteZoneConfig {
+		if err := newZoneConfig.Validate(); err != nil {
 			return pgerror.Newf(
 				pgcode.CheckViolation,
 				"could not validate zone config: %v",
 				err,
 			)
 		}
-		if err := zoneConfig.ValidateTandemFields(); err != nil {
+		if err := newZoneConfig.ValidateTandemFields(); err != nil {
 			return pgerror.Newf(
 				pgcode.CheckViolation,
 				"could not validate zone config: %v",
@@ -588,13 +626,13 @@ func ApplyZoneConfigForMultiRegionTable(
 			txn,
 			tableID,
 			table,
-			&zoneConfig,
+			&newZoneConfig,
 			execCfg,
 			hasNewSubzones,
 		); err != nil {
 			return err
 		}
-	} else if zoneRaw != nil {
+	} else if deleteZoneConfig {
 		// Delete the zone configuration if it exists but the new zone config is blank.
 		if _, err = execCfg.InternalExecutor.Exec(
 			ctx,

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -2436,8 +2436,20 @@ func (sc *SchemaChanger) applyZoneConfigChangeForMutation(
 				switch lcSwap.NewLocalityConfig.Locality.(type) {
 				case *descpb.TableDescriptor_LocalityConfig_Global_,
 					*descpb.TableDescriptor_LocalityConfig_RegionalByTable_:
-					// Nothing to do here. The table re-writing the locality config is all
-					// that is required.
+					// The table re-writing the LocalityConfig does most of the work for
+					// us here, but if we're coming from REGIONAL BY ROW, it's also
+					// necessary to drop the zone configurations on the index partitions.
+					if lcSwap.OldLocalityConfig.GetRegionalByRow() != nil {
+						opts = append(
+							opts,
+							dropZoneConfigsForMultiRegionIndexes(
+								append(
+									[]descpb.IndexID{pkSwap.OldPrimaryIndexId},
+									pkSwap.OldIndexes...,
+								)...,
+							),
+						)
+					}
 				case *descpb.TableDescriptor_LocalityConfig_RegionalByRow_:
 					// Apply new zone configurations for all newly partitioned indexes.
 					opts = append(

--- a/pkg/sql/set_zone_config.go
+++ b/pkg/sql/set_zone_config.go
@@ -972,14 +972,29 @@ func RemoveIndexZoneConfigs(
 		return nil
 	}
 
+	// Look through all of the subzones and determine if we need to remove any
+	// of them. We only want to rewrite the zone config below if there's actual
+	// work to be done here.
+	zcRewriteNecessary := false
 	for _, indexDesc := range indexDescs {
-		zone.DeleteIndexSubzones(uint32(indexDesc.ID))
+		for _, s := range zone.Subzones {
+			if s.IndexID == uint32(indexDesc.ID) {
+				// We've found an subzone that matches the given indexID. Delete all of
+				// this index's subzones and move on to the next index.
+				zone.DeleteIndexSubzones(uint32(indexDesc.ID))
+				zcRewriteNecessary = true
+				break
+			}
+		}
 	}
 
-	// Ignore CCL required error to allow schema change to progress.
-	_, err = writeZoneConfig(ctx, txn, tableDesc.GetID(), tableDesc, zone, execCfg, false /* hasNewSubzones */)
-	if err != nil && !sqlerrors.IsCCLRequiredError(err) {
-		return err
+	if zcRewriteNecessary {
+		// Ignore CCL required error to allow schema change to progress.
+		_, err = writeZoneConfig(ctx, txn, tableDesc.GetID(), tableDesc, zone, execCfg, false /* hasNewSubzones */)
+		if err != nil && !sqlerrors.IsCCLRequiredError(err) {
+			return err
+		}
 	}
+
 	return nil
 }

--- a/pkg/sql/type_change.go
+++ b/pkg/sql/type_change.go
@@ -500,6 +500,7 @@ func repartitionRegionalByRowTables(
 		tableDesc, err := localPlanner.Descriptors().GetMutableTableByID(
 			ctx, txn, tbID, tree.ObjectLookupFlags{
 				CommonLookupFlags: tree.CommonLookupFlags{
+					AvoidCached:    true,
 					Required:       true,
 					IncludeDropped: true,
 				},

